### ISSUE TITLE
["bug" linkage] can't resolve "__cdecl LightGBM::Tree::~Tree(void)" ...

### ISSUE THREAD

**Issue by lcsdavid** [microsoft/LightGBM#3412]
## ["bug" linkage] can't resolve "__cdecl LightGBM::Tree::~Tree(void)" ... on C++ library

For some reasons I need to interact with trees `LightGBM::Tree` more than it supposed to be as this class is not exported within LightGBM's DLL.
It results to a failure on linkage (perfectly normal) when trying to destroy those trees in this sample code:
``` 
std::vector<std::unique_ptr<LightGBM::Tree>> models = /* no matter where this come from. */;  

models.pop_back(); /* <-- will trigger fail on linkage because there is no symbol in DLL 'can't resolve "__cdecl LightGBM::Tree::~Tree(void)"' */
```

The easy solution for this is too put the destructor definition with his declaration, because destructor is trivial and therefore we can default it.
Solution would be in <tree.h>:
`~Tree() noexcept = default;`
It resolve this linkage since the definiton is in header.

(That's why I put it as a "bug" because it is not really one).

May I write a pull request ?

Best regards,

Edit: clarified issue

**Comment by jameslamb** [comment]
thanks for using `LightGBM` @lcsdavid ! I don't understand the issue from your description, but it might be clearer if we see your proposed change.

Please submit a pull request, thanks for contributing!

**Comment by StrikerRUS** [comment]
Fixed via #3417.
